bot/cogs/watchchannels/watchchannel.py

```diff
@@ -0,0 +1,359 @@
+
+
+def proxy_user(user_id: str) -> Object:
+    try:
+        user_id = int(user_id)
+    except ValueError:
+        raise BadArgument
+    user = Object(user_id)
+    user.mention = user.id
+    user.display_name = f"<@{user.id}>"
+    user.avatar_url_as = lambda static_format: None
+    user.bot = False
+    return user
+
+
+class WatchChannel(ABC):
+    """
+    Base class for WatchChannels
+
+    Abstracts the basic functionality for watchchannels in
+    a granular manner to allow for easy overwritting of
+    methods in the child class.
+    """
+
+    @abstractmethod
+    def __init__(self, bot: Bot, destination, webhook_id, api_endpoint, api_default_params, logger) -> None:
+        """
+        abstractmethod for __init__ which should still be called with super().
+
+        Note: Some of the attributes below need to be overwritten in the
+        __init__ of the child after the super().__init__(*args, **kwargs)
+        call.
+        """
+        self.bot = bot
+
+        self.destination = destination  # E.g., Channels.big_brother_logs
+        self.webhook_id = webhook_id  # E.g.,  Webhooks.big_brother
+        self.api_endpoint = api_endpoint  # E.g., 'bot/infractions'
+        self.api_default_params = api_default_params  # E.g., {'active': 'true', 'type': 'watch'}
+        self.log = logger  # Logger of the child cog for a correct name in the logs
+
+        # These attributes can be left as they are in the child class
+        self._consume_task = None
+        self.watched_users = defaultdict(dict)
+        self.message_queue = defaultdict(lambda: defaultdict(deque))
+        self.consumption_queue = {}
+        self.retries = 5
+        self.retry_delay = 10
+        self.channel = None
+        self.webhook = None
+        self.message_history = [None, None, 0]
+
+        self._start = self.bot.loop.create_task(self.start_watchchannel())
+
+    @property
+    def consuming_messages(self) -> bool:
+        """Checks if a consumption task is currently running."""
+        if self._consume_task is None:
+            return False
+
+        if self._consume_task.done():
+            exc = self._consume_task.exception()
+            if exc:
+                self.log.exception(
+                    f"{self.__class__.__name__} consume task has failed with:",
+                    exc_info=exc
+                )
+            return False
+
+        return True
+
+    async def start_watchchannel(self) -> None:
+        """Retrieves watched users from the API."""
+        await self.bot.wait_until_ready()
+
+        if await self.initialize_channel() and await self.fetch_user_cache():
+            self.log.trace(f"Started the {self.__class__.__name__} WatchChannel")
+        else:
+            self.log.error(f"Failed to start the {self.__class__.__name__} WatchChannel")
+
+            # Let's try again in a minute.
+            await asyncio.sleep(60)
+            self._start = self.bot.loop.create_task(self.start_watchchannel())
```

Comment:
I've restructured this part completely.

The TextChannel is the only part that actually needs retries, since the `fetch_*` methods of discord.py already have internal retry logic that takes the type of failure into account (retrying if the API is unreachable, not bothering again if there's a permissions problem or similar). (Note: the `get_webhook_info` is such a `fetch_*` method and has been renamed in newer versions to reflect that.)

I've removed the separate `initialize_channel` function, since it's only called when we start the watch channel anyway and moved the code over to `start_watchchannel`. I've only wrapped the `get_channel` in the retry logic, so we only have to replace those lines of code after we update to a newer version d.py.

If either the TextChannel or Webhook instances fail to be fetched/retrieved, the entire cog becomes non-functional, so I've followed your suggestion and included logic to unload the cog. Basically, it will log it, then tries to use the `ModLog` to send an alert to the mod-log channel, and finally unloads itself. (If the `ModLog` alert fails, then something is entirely wrong with the bot and we probably have no way of sending an alert by other means.)

The final step in the start-up is retrieving the watched user list from our API. If that fails, the cogs stays "functional" (without watched users), so it won't be unloaded, but it will send an alert notifying that we failed to retrieve the watched users. If it's just a temporary hiccup in the API, a simple cache refresh command could then fix the issue. (This occasionally happens now in the old implementation.)

Reply:
Looks great now.